ring/ring.go

```diff
@@ -66,35 +65,147 @@ func (rt Type) MarshalJSON() ([]byte, error) {
+
+	return &cr, nil
+}
+
+// StandardRing returns the standard ring of the receiver ring.
+// If `r.Type()==Standard`, then the method returns the receiver.
+// if `r.Type()==ConjugateInvariant`, then the method returns a ring with ring degree 2N.
+// The returned Ring is a shallow copy of the receiver.
+func (r *Ring) StandardRing() (*Ring, error) {
+	if r.Type() == Standard {
+		return r, nil
+	}
 
-	//NTT Parameters
-	NthRoot    uint64
-	PsiMont    []uint64 //2N-th primitive root in Montgomery form
-	PsiInvMont []uint64 //2N-th inverse primitive root in Montgomery form
+	sr := *r
+	sr.Tables = make([]*Table, len(r.Tables))
+	for i := range sr.Tables {
+		sr.Tables[i] = NewTable(r.N()<<1, r.Tables[i].Modulus)
+		sr.Tables[i].Factors = r.Tables[i].Factors
+		if err := r.Tables[i].GenNTTParams(uint64(sr.N()) << 1); err != nil {
+			return nil, err
+		}
+	}
+
+	sr.NumberTheoreticTransformer = NumberTheoreticTransformerStandard{}
+
+	return &sr, nil
+}
+
+// N returns the ring degree.
+func (r *Ring) N() int {
+	return r.Tables[0].N
+}
+
+// NthRoot returns the multiplicative order of the primitive root.
+func (r *Ring) NthRoot() uint64 {
+	return r.Tables[0].NthRoot
+}
+
+// NbModuli returns the number of primes in the RNS basis of the ring.
+func (r *Ring) NbModuli() int {
+	return len(r.Tables)
+}
 
-	NttPsi    [][]uint64 //powers of the inverse of the 2N-th primitive root in Montgomery form (in bit-reversed order)
-	NttPsiInv [][]uint64 //powers of the inverse of the 2N-th primitive root in Montgomery form (in bit-reversed order)
-	NttNInv   []uint64   //[N^-1] mod Qi in Montgomery form
+// Level returns the level of the current ring.
+func (r *Ring) Level() int {
+	return r.level
+}
+
+// AtLevel returns a shallowcopy of the target ring that operates at the target level.
+func (r *Ring) AtLevel(level int) *Ring {
+
+	if level < 0 {
+		panic("level cannot be negative")
+	}
+
+	if level > r.MaxLevel() {
+		panic("level cannot be larger than max level")
+	}
+
+	return &Ring{
+		NumberTheoreticTransformer: r.NumberTheoreticTransformer,
+		Tables:                     r.Tables,
+		ModulusAtLevel:             r.ModulusAtLevel,
+		RescaleParams:              r.RescaleParams,
+		level:                      level,
+	}
+}
+
+// MaxLevel returns the maximum level allowed by the ring (#NbModuli -1).
+func (r *Ring) MaxLevel() int {
+	return r.NbModuli() - 1
+}
+
+// Moduli returns the list of primes in the modulus chain.
+func (r *Ring) Moduli() (moduli []uint64) {
+	moduli = make([]uint64, len(r.Tables))
+	for i := range r.Tables {
+		moduli[i] = r.Tables[i].Modulus
+	}
+
+	return
+}
+
+// Modulus returns the modulus of the target ring at the currently
+// set level in *big.Int.
+func (r *Ring) Modulus() *big.Int {
+	return r.ModulusAtLevel[r.level]
+}
+
+// MRedParams returns the concatenation of the Montgomery parameters
+// of the target ring.
+func (r *Ring) MRedParams() (mredparams []uint64) {
```

Comment:
It might be possible to avoid this API point (and slice allocation) with only a slight modification of the RNS decomposition and reconstruction methods (they end up looping over the levels/MRedParams slice).

Reply:
It is more efficient to allocate a new slice with just the primes, than to loop over the subrings and their retrieve their individual primes.